design-proposals/flavors-and-preferences.md

```diff
@@ -0,0 +1,492 @@
+
+For example, if a user selects a `VirtualMachinePreference` that requests a `preferredDiskBus` of `virtio` but then sets a disk bus of `SATA` for one or more disk devices within the `VirtualMachine` the supplied `preferredDiskBus` preference will not be applied to these disks. Any remaining disks that do not have a disk bus defined will however use the `preferredDiskBus` preference of `virtio`.
+
+## Implementation Phases
+
+* Remove [`VirtualMachineFlavorProfile`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavorprofile) from [`VirtualMachineFlavor`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavor) and [`VirtualMachineClusterFlavor`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineclusterflavor).
+
+* Introduce a new `Spec` attribute and new `VirtualMachineFlavorSpec` type  itself with `CPU`, `Memory`, `GPUs` and `HostDevices` attributes to [`VirtualMachineFlavor`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavor) and [`VirtualMachineClusterFlavor`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineclusterflavor).
+
+* Introduce new `VirtualMachinePreference` and `VirtualMachineClusterPreference` CRDs to the Flavors API and `VirtualMachinePreferenceSpec` type to define guestOS related preferences such as preferred device buses and models.
+
+* Allow for `CPU` preferred topology policy to be defined within `VirtualMachinePreferenceSpec` through a `PreferredTopology` attribute.
+
+* Extend [`VirtualMachineSpec`](http://kubevirt.io/api-reference/main/definitions.html#_v1_virtualmachinespec) to allow an optional reference to `VirtualMachinePreference` or `VirtualMachineClusterPreference`.
+
+## [`DomainSpec`](http://kubevirt.io/api-reference/main/definitions.html#_v1_domainspec) mappings to `VirtualMachinePreferenceSpec` and `VirtualMachineFlavorSpec`
+
+The following basic rules are applied when deciding which attributes of the [`DomainSpec`](http://kubevirt.io/api-reference/main/definitions.html#_v1_domainspec) will be mapped to the new `VirtualMachinePreferenceSpec` and `VirtualMachineFlavorSpec` types:
+
+* Resource, schedulable and performance related attributes of the DomainSpec reside within `VirtualMachineFlavorSpec`
+
+* Guest OS related preferences reside within `VirtualMachinePreferenceSpec`
+
+For example, the number of and topology of vCPUs presented to the guest are defined within `VirtualMachineFlavorSpec` but the preferred topology of these vCPUs resides within  `VirtualMachinePreferenceSpec`.
+
+* [DomainSpec](http://kubevirt.io/api-reference/main/definitions.html#_v1_domainspec)
+  * [CPU](http://kubevirt.io/api-reference/main/definitions.html#_v1_cpu)
+    * cores 
+      * `VirtualMachineFlavorSpec.CPU.Guest` 
+        * Number of vCPUs to expose to the guest
+      * `VirtualMachinePreferenceSpec.CPU.PreferredTopology`
```

Comment:
I suppose this should go to flavor as well, reasoning: it's a hardware resource we expose.

Reply:
I had a conversation with @vladikr (keep me honest here please) about this a while ago and at the time I was advocating for the same given that's what we also did in OpenStack with extra specs such as [`hw:cpu_sockets`](https://docs.openstack.org/nova/latest/configuration/extra-specs.html#hw:cpu_sockets), [`hw:cpu_cores`](https://docs.openstack.org/nova/latest/configuration/extra-specs.html#hw:cpu_cores) and [`hw:cpu_threads`](https://docs.openstack.org/nova/latest/configuration/extra-specs.html#hw:cpu_threads).

I think the only reason we went  with PreferredTopology in the VirtualMachinePreferenceSpec was due to the Windows use case that keeps coming up where it prefers sockets over cores. So having it as a preference removes possible duplication from the flavors.